spec/lib/services/workflow-api-service-spec.js

```diff
@@ -0,0 +1,187 @@
+            expect(workflowApiService.createActiveGraph).to.have.been.calledOnce;
+            expect(workflowApiService.createActiveGraph).to.have.been.calledWith(
+                graphDefinition, { test: 1 }, { target: 'testnodeid', test: 2 }, 'test'
+            );
+            expect(workflowApiService.runTaskGraph).to.have.been.calledOnce;
+            expect(workflowApiService.runTaskGraph)
+                .to.have.been.calledWith(graph.instanceId, 'test');
+        });
+    });
+
+    it('should not create a graph against a node if there is an existing one active', function () {
+        workflowApiService.findGraphDefinitionByName.resolves(graphDefinition);
+        store.findActiveGraphForTarget.resolves({});
+
+        return expect(
+            workflowApiService.createAndRunGraph({
+                name: 'Graph.Test',
+                options: { test: 1 },
+                context: { test: 2 },
+                domain: 'test'
+            }, 'testnodeid')
+        ).to.be.rejectedWith(/Unable to run multiple task graphs against a single target/);
+    });
+
+    it('should return a NotFoundError if the node was not found', function () {
+        waterline.nodes.needByIdentifier.rejects(new Errors.NotFoundError('Not Found'));
+        return expect(workflowApiService.createAndRunGraph({}, 'testnodeid'))
+            .to.be.rejectedWith(Errors.NotFoundError);
+    });
+
+    it('should return a BadRequestError on a graph creation/validation failure', function () {
+        workflowApiService.createActiveGraph.restore();
+        workflowApiService.findGraphDefinitionByName.resolves({
+            tasks: [
+                { label: 'duplicate' },
+                { label: 'duplicate' }
+            ]
+        });
+
+        return expect(workflowApiService.createAndRunGraph({}, 'testnodeid'))
+            .to.be.rejectedWith(Errors.BadRequestError,
+                /The task label \'duplicate\' is used more than once/);
+    });
+
+    it('should return a NotFoundError if the node was not found', function () {
+        store.findActiveGraphForTarget.rejects(new Errors.NotFoundError('Not Found'));
+        return expect(workflowApiService.findActiveGraphForTarget('testnodeid'))
+            .to.be.rejectedWith(Errors.NotFoundError);
+    });
+
+    it('should persist a graph definition', function () {
+        store.persistGraphDefinition.resolves();
+        this.sandbox.stub(workflowApiService, 'createGraph').resolves();
+        return workflowApiService.defineTaskGraph(graphDefinition)
+        .then(function() {
+            expect(store.persistGraphDefinition).to.have.been.calledOnce;
+            expect(store.persistGraphDefinition).to.have.been.calledWith(graphDefinition);
+            expect(workflowApiService.createGraph).to.have.been.calledOnce;
+            expect(workflowApiService.createGraph).to.have.been.calledWith(graphDefinition);
+        });
+    });
+
+    it('should validate a graph definition', function () {
```

Comment:
Primarily what I wanted to test here was just that a validation error would bubble up. I do a full exercising of all the validation failure cases in these two sections of tests within on-core:

https://github.com/benbp/on-core/blob/taskgraph-improvements/spec/lib/workflow/task-graph-spec.js#L97-L200
https://github.com/benbp/on-core/blob/taskgraph-improvements/spec/lib/workflow/task-graph-traversal-spec.js#L19-L110